### ISSUE TITLE
Complete statement doesn't complete the statement

### ISSUE THREAD

**Issue by athomas** [dart-lang/sdk#38795]
Given this code (taken from a Flutter tutorial):

```dart
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    // ···
  }

  Column _buildButtonColumn(Color color, IconData icon, String label) {
    return Column(
      mainAxisSize: MainAxisSize.min,
      mainAxisAlignment: MainAxisAlignment.center,
      children: [
        Icon(icon, color: color),
        Container(
          margin: const EdgeInsets.only(top: 8),
          child: Text(
            label,
            style: TextStyle(
              fontSize: 12,
              fontWeight: FontWeight.w400,
              color: color,
```

I expected vscode's "Dart: Complete statement" to insert this for me:
```dart
            ),
          ),
        ),
      ],
    );
  }
}
```
Or at least:
```
            ),
          ),
        ),
      ],
    );
```

Instead, something weird gets inserted (just the last two lines):
```
              color: color);,,
    ]              
```

For a simpler case:
```
class MyApp {

  dynamic f() {
    return [
      {
        "key": [
          {
```
it inserts a comma and a closing bracket:
```
class MyApp {

  dynamic f() {
    return [
      {
        "key": [
          {,
    ]
```

Dart version is 2.6.0-dev.4.0. @DanTup

**Comment by DanTup** [comment]
The Complete Statement command was not brought over to LSP, so it hasn't been available for some time in VS Code (unless you explicitly turn off LSP).

@athomas @bwilkerson is there any feeling for how desirable/used this feature was? We could support it as a custom LSP command if we think it's worthwhile (and then fix issues like this one).

**Comment by bwilkerson** [comment]
No, I don't know. Adding some folks who might have an opinion.

@jacob314 @mit-mit

**Comment by mit-mit** [comment]
I have only anecdotal evidence that lots of developers find it hard to "complete the parentheses" for deep widget trees, so if this could help with that, I think it would be valuable.

**Comment by bwilkerson** [comment]
I wonder how discoverable that operation was, and whether it might not be better, at least in LSP, to simply include such completions in the normal code completion response.

**Comment by DanTup** [comment]
I suspect it was not very discoverable at all (if you did discover it and wanted to use it, you'd likely also need to set up a keybinding for it).

> whether it might not be better, at least in LSP, to simply include such completions in the normal code completion response.

That sounds like an interesting idea (and probably not difficult to add). Although, I do wonder what the common paths are to people ending up with mismatched braces etc.. By default, VS Code inserts a closing brace/parent when you type the opening one, so I would suspect the most common cases of mismatched braces/parens are when people are moving code around or copy/pasting rather than writing new statements, in which case I'm not sure how often this would help (if I understand correctly, it can only finish off a statement and not rewrite the closing symbols to make things valid).

If so, perhaps a better fix for that specific issue is a quick-fix on the diagnostics related to extra/missing/incorrect closing symbols. That's probably more difficult though, and there may be cases where the fix is ambigious.

**Comment by bwilkerson** [comment]
> ... I do wonder what the common paths are to people ending up with mismatched braces ...

I agree with your assessment. I suspect that copy/paste is the most common source of mismatched braces.

> ... perhaps a better fix for that specific issue is a quick-fix ...

Yes, the fix is often ambiguous in the sense that there can be multiple valid places to add the closing token. That's one of the reasons why recovering from such errors is so difficult to do well.

Personally I think that the best fix for that issue is the "intelligent copy/paste" support we prototyped in IntelliJ several years ago. That would allow us to see what the original code looked like before it was copied and to sometimes use that information to fix things up when pasting. I think that could be done using the experimental `documentPaste` request, if it is adopted.

Although, that doesn't preclude having both features if that would improve the situation even more.
